### PR TITLE
Updated HCS Acceptance tests for easy running by devops

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
DevOps is looking to automate usage of our acceptance tests.
To assist this the PR includes changes to make the tests more configurable but also more conservative on assessments.

- Added retry logic using spring-retry annotations on multiple service calls to recover from situations where the transaction for topic creation or message publishing may not have hit the db yet.
- Added background message publishing to encourage service file closing and reducing wait time on message in environments with low tps
- Updated Readme file
- Updated properties config with new emitBackgroundMessages, existingTopicNum. Changed timeout to duration
- Improved some of the logging.

**Special notes for your reviewer**:
Tests can be configured to be more aggressive in regards to latency. Defaults just reduce necessary configuration but still ensure some good traffic to nodes.

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by apeksharma** [comment]
why add subscribeDelay ?
- It's shouldn't be needed. And at the very least, there should be one test which subscribes immediately because that's a valid user pattern we should test for.
- How does it make it easier for devops to run the acceptance tests? I believe they don't run it right now since the steps are: update repo -> clean install -> run the tests, which is not what devops should be doing. They should have one click/command way to run it. So until we give that to devops, the onus is on us to run them. Not saying we should make it happen here. But just trying to understand these changes.

**Comment by Nana-EC** [comment]
> why add subscribeDelay ?
> 
> * It's shouldn't be needed. And at the very least, there should be one test which subscribes immediately because that's a valid user pattern we should test for.
> * How does it make it easier for devops to run the acceptance tests? I believe they don't run it right now since the steps are: update repo -> clean install -> run the tests, which is not what devops should be doing. They should have one click/command way to run it. So until we give that to devops, the onus is on us to run them. Not saying we should make it happen here. But just trying to understand these changes.

Actively explored a scenario with them today in regards to resolving the proxy issue.
Their plan is to containerize this and use the configs to point to any env to run acceptance tests as part of deployment.

The subscribeDelay is needed because occasionally the file isn't closed from the service side and as such the transactions submitted don' always make it to the mirror node as quickly as expected in non testnet and Mainnet environments. This gives the option to be able to be a bit more conservative with latency assessments.
Agin these are all configuration and do not have to be used. I've only preset it to dev ops usage and like I said you can be more aggressive with your settings.

Feel free to chat with Gabriel if you want more context for his plans on this but it seems fair as we don't want devops having to debug these test sin the edge cases of non production environments.

Let me know if that clarifies things.

**Comment by apeksharma** [comment]
I see.

> ...such the transactions submitted don' always make it to the mirror node as quickly as expected...
To address that case, we can wait for messages with a timeout (just like java tests have timeouts when they are testing for any 'eventually X happens' logic).

`subscribeDelay` is a pattern special to non-mainnet/non-testnet envs.
'Waiting for messages with timeout' is a pattern which'll be true for all envs.
We should build our test suite to adapt the latter since it's right way to adapt the issue mentioned in previous comment.

**Comment by Nana-EC** [comment]
> I see.
> 
> > ...such the transactions submitted don' always make it to the mirror node as quickly as expected...
> 
> To address that case, we can wait for messages with a timeout (just like java tests have timeouts when they are testing for any 'eventually X happens' logic).
> 
> `subscribeDelay` is a pattern special to non-mainnet/non-testnet envs.
> 'Waiting for messages with timeout' is a pattern which'll be true for all envs.
> We should build our test suite to adapt the latter since it's right way to adapt the issue mentioned in previous comment.

There's already a wait for messages property that's been in use from the beginning - messageTimeout.
However this does not affect the topic existing as the service will just error out if the topic doesn't exist. Messages are streamed so as along was the channel is open you can wait a little.

Again subscribeDelay is an additional property for use in non production ends as you've said. It's conferrable so it can be made 0 by us who understand the intricacies of the whole flow. It's just defaulted to a non zero value for devops so they can just run with minimal config.

Let me know if I've further clarified things and if there is a follow up specific ask on this PR.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 1]
```diff
@@ -38,16 +38,19 @@
...
         log.debug("Creating Mirror Node client for {}", mirrorNodeAddress);
         mirrorClient = new MirrorClient(Objects.requireNonNull(mirrorNodeAddress));
         messageTimeout = acceptanceTestProperties.getMessageTimeout();
+        subscribeDelay = acceptanceTestProperties.getSubscribeDelay() * 1000;
     }
 
     public SubscriptionResponse subscribeToTopic(MirrorConsensusTopicQuery mirrorConsensusTopicQuery) throws Throwable {
-        log.debug("Subscribing to topic");
+        log.debug("Subscribing to topic. Subscribe is delayed by {} ms", subscribeDelay);
+        Thread.sleep(subscribeDelay, 0);
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: I think instead of a subscribe delay this should be a subscribe timeout. Subscribe delay is a slow and brittle approach that always waits that amount of time before subscribing. Brittle because if it's not an appropriate value for that environment it will fail. A subscribe timeout is a smarter approach that would retry subscribe every 1s up to the timeout when a topic not found is returned, potentially returning quicker than the delay.

I understand you have message timeout, but that should probably be adjusted to start after subscription is successful. This may have been what Appy was getting at.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java [hunk 2]
```diff
@@ -44,4 +45,8 @@
     private String operatorKey;
     @Min(0)
     private int messageTimeout;
+    @Min(0)
+    private int subscribeDelay;
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java]: These timeouts/delays should be a Duration

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 3]
```diff
@@ -38,16 +38,19 @@
 public class MirrorNodeClient {
     private final MirrorClient mirrorClient;
     private final int messageTimeout;
+    private final int subscribeDelay;
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: Why are we duplicately storing these instead of just storing the AcceptanceTestProperties?

File: hedera-mirror-test/src/test/resources/features/hcs/hcs.feature [hunk 4]
```diff
@@ -5,23 +5,23 @@ Feature: HCS Base Coverage Feature
...
         And I publish and verify <numMessages> messages sent
-        When I provide a number of messages <numMessages> I want to receive
+        When I provide a number of messages <expectedCount> I want to receive
         And I subscribe with a filter to retrieve messages
         Then the network should successfully observe these messages
         Examples:
-            | numMessages |
-            | 100         |
+            | numMessages | expectedCount |
+            | 15          | 10            |
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/resources/features/hcs/hcs.feature]: > Updated .feature files to specify an expectedCount of messages which can differ from the published amount. This allows for environments where transaction volume is small and files aren't closed as often.

As a non-programmer, I should be able to read these feature files and understand them. These changes makes it non-obvious and confusing. A better approach would be to revert to the old feature file with a single numMessages, then publish numMessages and every second after publish one more message until timeout. The additional messages are just for closing the record file and can be ignored and you just verify you received at least numMessages.

Comment (Nana-EC) [reply hedera-mirror-test/src/test/resources/features/hcs/hcs.feature]: I see what you're saying. However, I did also observe that 1 message extra isn't always enough to close the record file.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: I should store the object over the individual primitives. Will fix.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: Yeah, Appy and I agreed on this. I was hoping to do retry in a different PR within the spring but wanted to get this in and focus on a few other things earlier on.

Let me know if this is blocking or if I can address this later as it works for now.

File: hedera-mirror-test/src/test/resources/features/hcs/hcs.feature (hunk 4 above)
Comment (steven-sheehy) [reply hedera-mirror-test/src/test/resources/features/hcs/hcs.feature]: Not one extra message, one message per second until timeout.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 1 above)
Comment (steven-sheehy) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: I'm okay to go with this approach and replace it with subscribeTimeout in a follow up.

**Comment by Nana-EC** [comment]
> > I see.
> > > ...such the transactions submitted don' always make it to the mirror node as quickly as expected...
> > 
> > 
> > To address that case, we can wait for messages with a timeout (just like java tests have timeouts when they are testing for any 'eventually X happens' logic).
> > `subscribeDelay` is a pattern special to non-mainnet/non-testnet envs.
> > 'Waiting for messages with timeout' is a pattern which'll be true for all envs.
> > We should build our test suite to adapt the latter since it's right way to adapt the issue mentioned in previous comment.
> 
> There's already a wait for messages property that's been in use from the beginning - messageTimeout.
> However this does not affect the topic existing as the service will just error out if the topic doesn't exist. Messages are streamed so as along was the channel is open you can wait a little.
> 
> Again subscribeDelay is an additional property for use in non production ends as you've said. It's conferrable so it can be made 0 by us who understand the intricacies of the whole flow. It's just defaulted to a non zero value for devops so they can just run with minimal config.
> 
> Let me know if I've further clarified things and if there is a follow up specific ask on this PR.

@apeksharma I addressed your original feedback since I re prioritized my tasks. Removed the delay and added retry logic

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: Went ahead and implemented retry logic

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: removed

File: hedera-mirror-test/src/test/resources/features/hcs/hcs.feature (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/resources/features/hcs/hcs.feature]: Implemented this with a background thread that emits messages once every second as you noted

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java]: Fixed and removed one

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java [hunk 5]
```diff
@@ -198,17 +223,17 @@ public void verifySubscriptionChannelConnection() throws Throwable {
     }
 
     @When("I publish {int} batches of {int} messages every {long} milliseconds")
+    @Retryable(value = {StatusRuntimeException.class})
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java]: I don't think you can retry a batch, right? You might've sent one successfully then got an exception on second. It would then try to send the first one again. Retry should be on the single publish message if at all.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java [hunk 6]
```diff
@@ -312,4 +337,56 @@ public void closeClients() {
...
+     *
+     * @return SubscriptionResponse
+     * @throws InterruptedException
+     */
+    public SubscriptionResponse subscribeWithBackgroundMessageEmission() throws Throwable {
+        Thread backgroundMessageThread = null;
+
+        if (acceptanceProps.isEmitBackgroundMessages()) {
+            backgroundMessageThread = new Thread(topicClient
+                    .getBackgroundMessagePublisher(consensusTopicId, submitKey, (int) acceptanceProps
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java]: 3rd param is numBatches so why is a message timeout being passed?

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java [hunk 7]
```diff
@@ -312,4 +337,56 @@ public void closeClients() {
...
+     * environments with low traffic.
+     *
+     * @return SubscriptionResponse
+     * @throws InterruptedException
+     */
+    public SubscriptionResponse subscribeWithBackgroundMessageEmission() throws Throwable {
+        Thread backgroundMessageThread = null;
+
+        if (acceptanceProps.isEmitBackgroundMessages()) {
+            backgroundMessageThread = new Thread(topicClient
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java]: This is all  quite tightly coupled together. A feature should not know about threads, retrying or background production of messages. Would be better if topicClient handled it internally via one of its existing methods.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 8]
```diff
@@ -62,11 +65,12 @@ public SubscriptionResponse subscribeToTopic(MirrorConsensusTopicQuery mirrorCon
         return subscriptionResponse;
     }
 
+    @Retryable(value = {StatusRuntimeException.class})
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java:69]: Will look into this

File: hedera-mirror-test/src/test/resources/application-default.yml [hunk 9]
```diff
@@ -2,9 +2,12 @@ hedera:
...
+        emitBackgroundMessages: false
+        existingTopicNum:
+        messageTimeout: 20s
+        mirrorNodeAddress: hcs.testnet.mirrornode.hedera.com:5600
         nodeId: 0.0.3
         nodeAddress: testnet
-        mirrorNodeAddress: hcs.testnet.mirrornode.hedera.com:5600
         operatorId:
         operatorKey:
+        subscribeDelay: 2s
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-test/src/test/resources/application-default.yml:13]: Yes unused, meant to remove.
Not sure what you mean on limit here. 
Subscribes already have the messageTimeOut. They'll last as max the length of time the wait for messages.
Numbest of retries is already set in the Configuration file

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java [hunk 10]
```diff
@@ -42,6 +43,10 @@
     private String operatorId;
     @NotBlank
     private String operatorKey;
-    @Min(0)
-    private int messageTimeout;
+    @NotNull
+    private Duration messageTimeout;
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/AcceptanceTestProperties.java:47]: Will set.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java [hunk 11]
```diff
@@ -41,27 +44,32 @@
...
-
     @Autowired
     private TopicClient topicClient;
 
+    public TopicFeature(AcceptanceTestProperties acceptanceTestProperties) {
+        acceptanceProps = acceptanceTestProperties;
+    }
+
     @Given("I successfully create a new topic id")
+    @Retryable(value = {StatusRuntimeException.class})
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:72]: Initially the refuse stream also occurred on createTopic, so a retry on topic creation was valid where a new topic is needed - which is most tests.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:226]: Makes sense to retry a single emit. I'll update that

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:353]: I believe the recommendation was yours to emit a message every second for as long as we are listening for messages.
Therefore a 20 sec max timeout would mean emit at worst a message every seconds 20 times.
That's the logic here.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:352]: Yep agreed. This was part of the stuff I mentioned could be refactored but could wait.

File: hedera-mirror-test/src/test/resources/application-default.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-test/src/test/resources/application-default.yml:13]: There's no retries in config file. It seems to be hardcoded in `ClientConfiguration`.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:72]: Ok, so just need to check for UNAVAILABLE here then

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 6 above)
Comment (steven-sheehy) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:353]: Yeah, just confusing to have a numBatches be populated by a timeout. Would be better to use a ScheduledExecutorService in the future to run until the timeout:
 ```java
Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(() -> {}, 0, 1, TimeUnit.SECONDS).get(timeout, TimeUnit.Seconds);
```

File: hedera-mirror-test/src/test/resources/application-default.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-test/src/test/resources/application-default.yml:13]: Oh understood, you want to be able to configure the number of retries. Gotcha. Will do this

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java]: Overloaded method to retry single attempt to publishes messages not a batch.
As part of a larger refactor I'd have moved this into TopicClient but I'm hitting some issues and this works for now.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 12]
```diff
@@ -37,17 +40,18 @@
...
 
     public MirrorNodeClient(AcceptanceTestProperties acceptanceTestProperties) {
         String mirrorNodeAddress = acceptanceTestProperties.getMirrorNodeAddress();
         log.debug("Creating Mirror Node client for {}", mirrorNodeAddress);
         mirrorClient = new MirrorClient(Objects.requireNonNull(mirrorNodeAddress));
-        messageTimeout = acceptanceTestProperties.getMessageTimeout();
+        acceptanceProps = acceptanceTestProperties;
     }
 
+    @Retryable(value = {StatusRuntimeException.class}, exceptionExpression = "NOT_FOUND")
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java:52]: Don't think this will work. Did you try? It should be a spel expression that interrogates the object:
```java
@Retryable(include = {StatusRuntimeException.class}, exceptionExpression = "#{message.contains('this can be retried')}")`

@Retryable(include = {StatusRuntimeException.class}, exceptionExpression = "#{status == Status.NOT_FOUND}")`
```

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java [hunk 13]
```diff
@@ -201,14 +229,22 @@ public void verifySubscriptionChannelConnection() throws Throwable {
...
             Thread.sleep(milliSleep, 0);
+            publishTopicMessages(messageCount);
+            log.trace("Emitted {} message(s) in batch {} of {} potential batches. Will sleep {} ms until " +
+                    "next batch", messageCount, i + 1, numGroups, milliSleep);
         }
 
         messageSubscribeCount = numGroups * messageCount;
     }
 
+    @Retryable(value = {StatusRuntimeException.class})
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/TopicFeature.java:241]: I thought it was implied, but will state explicitly: exceptionExpression should be populated on all retryables
